File: src/client/console/libnxmc/policyEditor.cpp
/* $Id$ */
/* 
** NetXMS - Network Management System
** Portable management console
** Copyright (C) 2008 Victor Kirhenshtein
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
**/

#include "libnxmc.h"
#include "policyEditor.h"


//
// Cell renderer for object list
//

class nxGridCellObjListRenderer : public wxGridCellRenderer
{
public:
	virtual void Draw(wxGrid& grid, wxGridCellAttr& attr, wxDC& dc, const wxRect& rect, int row, int col, bool isSelected);
	virtual wxSize GetBestSize(wxGrid& grid, wxGridCellAttr& attr, wxDC& dc, int row, int col);
	virtual wxGridCellRenderer *Clone() const;
};

void nxGridCellObjListRenderer::Draw(wxGrid& grid, wxGridCellAttr& attr, wxDC& dc, const wxRect& rect, int row, int col, bool isSelected)
{
	dc.DrawText( _T("OBJ"), rect.x, rect.y);
}

wxSize nxGridCellObjListRenderer::GetBestSize(wxGrid& grid, wxGridCellAttr& attr, wxDC& dc, int row, int col)
{
	return wxSize(100, 40);
}

wxGridCellRenderer *nxGridCellObjListRenderer::Clone() const
{ 
	return new nxGridCellObjListRenderer; 
}


//
// Cell editor for object list
//

class nxGridCellObjListEditor : public wxGridCellEditor
{
public:
	nxGridCellObjListEditor() : wxGridCellEditor() { }
};


//
// Event table
//

BEGIN_EVENT_TABLE(nxPolicyEditor, wxGrid)
END_EVENT_TABLE()


//
// Constructor
//

nxPolicyEditor::nxPolicyEditor(wxWindow *parent, wxWindowID id)
			   : wxGrid(parent, id)
{
	RegisterDataType(_T("OBJECT_LIST"), new nxGridCellObjListRenderer, NULL);
	SetLabelBackgroundColour(wxColour(0, 0, 127));
	SetLabelTextColour(wxColour(255, 255, 255));
}


//
// Destructor
//

nxPolicyEditor::~nxPolicyEditor()
{
}


//
// Initializer
//

void nxPolicyEditor::Initialize(wxGridTableBase *table)
{
	SetTable(table, true, wxGrid::wxGridSelectRows);
}

File: src/java/netxms-eclipse/EPP/src/org/netxms/ui/eclipse/epp/views/PolicyEditor.java
/**
 * NetXMS - open source network management system
 * Copyright (C) 2003-2009 Victor Kirhenshtein
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package org.netxms.ui.eclipse.epp.views;

import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.IStatus;
import org.eclipse.core.runtime.Status;
import org.eclipse.core.runtime.jobs.Job;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.FillLayout;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.ExpandBar;
import org.eclipse.ui.part.ViewPart;
import org.eclipse.ui.progress.IWorkbenchSiteProgressService;
import org.eclipse.ui.progress.UIJob;
import org.netxms.client.NXCException;
import org.netxms.client.NXCSession;
import org.netxms.client.events.EventProcessingPolicy;
import org.netxms.ui.eclipse.epp.Activator;
import org.netxms.ui.eclipse.epp.widgets.RuleList;
import org.netxms.ui.eclipse.jobs.ConsoleJob;
import org.netxms.ui.eclipse.shared.NXMCSharedData;

/**
 * @author Victor
 *
 */
public class PolicyEditor extends ViewPart
{
	public static final String ID = "org.netxms.ui.eclipse.epp.view.policy_editor";
	public static final String JOB_FAMILY = "PolicyEditorJob";

	private NXCSession session;
	private boolean policyLocked = false;
	private EventProcessingPolicy policy; 
	private RuleList ruleList;
	
	/* (non-Javadoc)
	 * @see org.eclipse.ui.part.WorkbenchPart#createPartControl(org.eclipse.swt.widgets.Composite)
	 */
	@Override
	public void createPartControl(Composite parent)
	{
		session = NXMCSharedData.getInstance().getSession();

		parent.setLayout(new FillLayout());
		
		ruleList = new RuleList(parent, SWT.NONE);
		ConsoleJob job = new ConsoleJob("Open event processing policy", this, Activator.PLUGIN_ID, null) {
			@Override
			protected String getErrorMessage()
			{
				return "Cannot open event processing policy";
			}

			@Override
			protected void runInternal(IProgressMonitor monitor) throws Exception
			{
				policy = session.openEventProcessingPolicy();
				policyLocked = true;
				new UIJob("Fill rules list") {
					@Override
					public IStatus runInUIThread(IProgressMonitor monitor)
					{
						ruleList.setContent(policy.getRules());
						return Status.OK_STATUS;
					}
				}.schedule();
			}

			@Override
			protected void jobFailureHandler()
			{
				new UIJob("Close event processing policy editor")
				{
					@Override
					public IStatus runInUIThread(IProgressMonitor monitor)
					{
						PolicyEditor.this.getViewSite().getPage().hideView(PolicyEditor.this);
						return Status.OK_STATUS;
					}
				}.schedule();
			}
		};
		job.start();
	}

	/* (non-Javadoc)
	 * @see org.eclipse.ui.part.WorkbenchPart#setFocus()
	 */
	@Override
	public void setFocus()
	{
		// TODO Auto-generated method stub

	}

	/* (non-Javadoc)
	 * @see org.eclipse.ui.part.WorkbenchPart#dispose()
	 */
	@Override
	public void dispose()
	{
		if (policyLocked)
		{
			new Job("Close event processing policy")
			{
				@Override
				protected IStatus run(IProgressMonitor monitor)
				{
					IStatus status;

					try
					{
						status = Status.OK_STATUS;
					}
					catch(Exception e)
					{
						status = new Status(Status.ERROR, Activator.PLUGIN_ID,
								(e instanceof NXCException) ? ((NXCException) e).getErrorCode() : 0,
								"Cannot close event processing policy: " + e.getMessage(), null);
					}
					return status;
				}
			}.schedule();
		}
		super.dispose();
	}
}
